From 79449817aafa959a2e822e7adad3bd36cbe81d6e Mon Sep 17 00:00:00 2001
From: chandavath <rcrajagopal@habana.ai>
Date: Wed, 27 Mar 2024 07:24:38 +0200
Subject: [PATCH] [SW-180247] gaudi3: erom: issue slave die reset via spi gw

In eROM, d2d link is not up and we can't issue reset via d2d reg base.
So, issue slave die reset via SPI GW.

Added mirror api's to access slave register via SPI GW.

Replaced hardcoded value with proper macro.

Change-Id: I748835f478d1190c19d79aa9a68520ba81ea5de4
Signed-off-by: chandavath <rcrajagopal@habana.ai>
---

diff --git a/arch/gaudi3/common/reset.c b/arch/gaudi3/common/reset.c
index 80d0d80..3479259 100644
--- a/arch/gaudi3/common/reset.c
+++ b/arch/gaudi3/common/reset.c
@@ -20,12 +20,16 @@
 
 	regmap_rmwritel(D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_DIE_READY_IN), 0x0,
 					GLOBAL_CONF_DIE_READY_IN_IND_M);
-	/* Issue reset to slave die */
-	regmap_writel(0x1, D1_PSOC_RESET_CONF_REG(mmPSOC_RESET_CONF_SW_ALL_RST));
+
+	/* Issue reset to slave die via SPI GW */
+	mirror_writel(GLOBAL_CONF_DIE_READY_IN_IND_M,
+				D1_PSOC_RESET_CONF_REG(mmPSOC_RESET_CONF_SW_ALL_RST));
+
 	/* wait for slave die to reset no indication is required */
 	udelay(DELAY_1K_USEC);
 	/* Issue master die0 reset */
-	regmap_writel(0x1, D0_PSOC_RESET_CONF_REG(mmPSOC_RESET_CONF_SW_ALL_RST));
+	regmap_writel(GLOBAL_CONF_DIE_READY_IN_IND_M,
+				D0_PSOC_RESET_CONF_REG(mmPSOC_RESET_CONF_SW_ALL_RST));
 }
 
 void hl_dev_reset(void)
diff --git a/include/baremetal.h b/include/baremetal.h
index 4dcdef6..d40e4ed 100644
--- a/include/baremetal.h
+++ b/include/baremetal.h
@@ -149,6 +149,20 @@
 	regmap_writel(val, addr);
 }
 
+#ifdef HL_SPI_GW_CFG_BASE
+static inline void mirror_writel(u32 val, u32 offset)
+{
+	*((volatile u32 *)(HL_SPI_GW_CFG_BASE + offset)) = val;
+}
+
+static inline u32 mirror_readl(u32 offset)
+{
+	u32 reg = *((volatile u32 *)(HL_SPI_GW_CFG_BASE + offset));
+
+	return reg;
+}
+#endif
+
 u32 hl_skip_pcie_init(void);
 u32 hl_skip_fw_cfg(void);
 void hl_uart_init(void);
